File: App/DesignSystem/ISG/NovaSessionHost.swift
import Foundation

/// Local presentation identity supplied by the trusted Auth owner, never user metadata or a URL.
/// A token refresh retains sessionID; logout/re-login must adopt the new sessionID.
struct NovaSessionIdentity: Equatable {
    let userID: UUID
    let sessionID: UUID
}

/// The selected server context, reduced to the fields needed to invalidate UI work.
/// It is a cache boundary, never an authorization decision.
struct NovaWorkspaceSelection: Equatable {
    let workspaceID: UUID
    let membershipID: UUID
    let userID: UUID
    let kind: String
    let permissionRevision: Int64
    let workspaceVersion: Int64
    let canRead: Bool
    let canOperate: Bool

    var isStructurallyValid: Bool {
        ["personal", "osgb"].contains(kind) && permissionRevision >= 0 &&
        permissionRevision <= 9_007_199_254_740_991 && workspaceVersion >= 0 &&
        workspaceVersion <= 9_007_199_254_740_991
    }
}

enum NovaHostPhase: String { case signedOut, resolving, ready, failed }

struct NovaAvailabilityTicket: Equatable {
    fileprivate let id: UUID
    fileprivate let epoch: String
    fileprivate let identity: NovaSessionIdentity
    fileprivate let workspace: NovaWorkspaceSelection?
}

/// Opaque, in-memory value. Read through the CURRENT host, not a captured host snapshot.
struct NovaScopedValue<Value> {
    fileprivate let epoch: String
    fileprivate let value: Value
}

/// Value-state reducer. Serialize all access on the UI owner (MainActor); not server authorization.
/// No Auth SDK, billing inference, persistence, token parsing, company grants or network calls.
struct NovaSessionHost {
    private(set) var identity: NovaSessionIdentity?
    private(set) var workspace: NovaWorkspaceSelection?
    private(set) var phase: NovaHostPhase = .signedOut
    private(set) var navigation = NovaNavigationState(epoch: UUID().uuidString, available: [])
    private(set) var pending: NovaAvailabilityTicket?
    private let implemented: Set<NovaDestination>

    init(implemented: Set<NovaDestination>) { self.implemented = implemented }

    /// Duplicate Auth/token-refresh events do not erase paths or invalidate an in-flight load.
    mutating func adopt(_ next: NovaSessionIdentity?) {
        guard identity != next else { return }
        identity = next
        workspace = nil
        invalidate()
        phase = next == nil ? .signedOut : .resolving
    }

    /// Explicit permission revalidation hides old content and invalidates every old callback.
    /// A retry also supersedes its predecessor even when the user/session is unchanged.
    mutating func beginAvailabilityRefresh() -> NovaAvailabilityTicket? {
        guard let identity else { return nil }
        invalidate()
        phase = .resolving
        let ticket = NovaAvailabilityTicket(id: UUID(), epoch: navigation.epoch, identity: identity, workspace: workspace)
        pending = ticket
        return ticket
    }

    /// A workspace change is an account-scope boundary: hide the previous workspace
    /// immediately and require a new availability result before rendering content.
    mutating func beginWorkspaceSwitch(to next: NovaWorkspaceSelection) -> NovaAvailabilityTicket? {
        guard let identity, next.isStructurallyValid, next.canRead,
              next.userID == identity.userID, workspace != next else { return nil }
        workspace = next
        invalidate()
        phase = .resolving
        let ticket = NovaAvailabilityTicket(id: UUID(), epoch: navigation.epoch, identity: identity, workspace: next)
        pending = ticket
        return ticket
    }

    @discardableResult
    mutating func resolve(_ ticket: NovaAvailabilityTicket, ownerID: UUID, enabled: Set<NovaDestination>) -> Bool {
        guard matches(ticket) else { return false }
        pending = nil
        guard ownerID == identity?.userID else { phase = .failed; return false }
        navigation.updateAvailability(enabled.intersection(implemented), from: navigation.epoch)
        phase = .ready
        return true
    }

    @discardableResult
    mutating func fail(_ ticket: NovaAvailabilityTicket) -> Bool {
        guard matches(ticket) else { return false }
        pending = nil
        phase = .failed
        return true
    }

    mutating func apply(_ event: NovaNavigationEvent, from epoch: String) {
        guard phase == .ready else { return }
        navigation.apply(event, from: epoch)
    }

    /// SwiftUI Binding bridge. An old child cannot replace the current account or capability set.
    mutating func acceptNavigation(_ next: NovaNavigationState, from epoch: String) {
        guard isCurrent(epoch), next.epoch == navigation.epoch,
              next.available == navigation.available else { return }
        navigation = next
    }

    func isCurrent(_ epoch: String) -> Bool { phase == .ready && epoch == navigation.epoch }

    func isCurrent(_ epoch: String, workspaceID: UUID, permissionRevision: Int64) -> Bool {
        isCurrent(epoch) && workspace?.workspaceID == workspaceID &&
        workspace?.permissionRevision == permissionRevision
    }

    /// Does NOT implement latest-request-wins within an epoch: feature loaders must also compare
    /// their own request ID, company scope and query before publishing a response.
    func scope<Value>(_ value: Value, from epoch: String) -> NovaScopedValue<Value>? {
        guard isCurrent(epoch) else { return nil }
        return NovaScopedValue(epoch: epoch, value: value)
    }

    func value<Value>(from snapshot: NovaScopedValue<Value>?) -> Value? {
        guard let snapshot, isCurrent(snapshot.epoch) else { return nil }
        return snapshot.value
    }

    private func matches(_ ticket: NovaAvailabilityTicket) -> Bool {
        phase == .resolving && pending == ticket && identity == ticket.identity &&
        workspace == ticket.workspace && navigation.epoch == ticket.epoch
    }

    private mutating func invalidate() {
        pending = nil
        navigation = NovaNavigationState(epoch: UUID().uuidString, available: [])
    }
}

File: android/core/designsystem/src/main/kotlin/com/riskdetectedan/core/designsystem/isg/NovaSessionHost.kt
package com.riskdetectedan.core.designsystem.isg

import java.util.UUID

/** Trusted Auth-owner input, never profile metadata or URL fields; contains no credentials. */
data class NovaSessionIdentity(val userID: UUID, val sessionID: UUID)
data class NovaWorkspaceSelection(
    val workspaceID: UUID,
    val membershipID: UUID,
    val userID: UUID,
    val kind: String,
    val permissionRevision: Long,
    val workspaceVersion: Long,
    val canRead: Boolean,
    val canOperate: Boolean,
) {
    val isStructurallyValid: Boolean get() = kind in setOf("personal", "osgb") &&
        permissionRevision in 0..9007199254740991L && workspaceVersion in 0..9007199254740991L
}
enum class NovaHostPhase { signedOut, resolving, ready, failed }

// Opaque tokens: only the host issues these; do not serialize them or log identity fields.
class NovaAvailabilityTicket internal constructor(internal val epoch: String, internal val identity: NovaSessionIdentity,
    internal val workspace: NovaWorkspaceSelection?)
class NovaScopedValue<Value> internal constructor(internal val epoch: String, internal val value: Value)

/** Immutable presentation owner; assign results to the CURRENT UI state on the UI thread.
 * No Auth/billing/role inference, persistence, network or server authorization is provided.
 */
class NovaSessionHost private constructor(
    val identity: NovaSessionIdentity?,
    val workspace: NovaWorkspaceSelection?,
    val phase: NovaHostPhase,
    val navigation: NovaNavigationState,
    val pending: NovaAvailabilityTicket?,
    private val implemented: Set<NovaDestination>,
) {
    constructor(implemented: Set<NovaDestination>) : this(null, null, NovaHostPhase.signedOut,
        NovaNavigationState(UUID.randomUUID().toString(), emptySet()), null, implemented.toSet())

    /** Same session / token refresh is a no-op, including during an in-flight request. */
    fun adopt(next: NovaSessionIdentity?): NovaSessionHost = if (identity == next) this else
        NovaSessionHost(next, null, if (next == null) NovaHostPhase.signedOut else NovaHostPhase.resolving,
            freshNavigation(), null, implemented)

    /** Revalidation and retries invalidate old content/callbacks even for the same session. */
    fun beginAvailabilityRefresh(): NovaSessionHost {
        val actor = identity ?: return this
        val next = freshNavigation()
        return NovaSessionHost(actor, workspace, NovaHostPhase.resolving, next,
            NovaAvailabilityTicket(next.epoch, actor, workspace), implemented)
    }

    /** Workspace selection is a cache boundary and still requires server availability resolution. */
    fun beginWorkspaceSwitch(nextWorkspace: NovaWorkspaceSelection): NovaSessionHost {
        val actor = identity ?: return this
        if (!nextWorkspace.isStructurallyValid || !nextWorkspace.canRead || nextWorkspace.userID != actor.userID ||
            workspace == nextWorkspace) return this
        val next = freshNavigation()
        return NovaSessionHost(actor, nextWorkspace, NovaHostPhase.resolving, next,
            NovaAvailabilityTicket(next.epoch, actor, nextWorkspace), implemented)
    }

    fun resolve(ticket: NovaAvailabilityTicket, ownerID: UUID, enabled: Set<NovaDestination>): NovaSessionHost {
        if (!matches(ticket)) return this
        if (ownerID != identity?.userID) return NovaSessionHost(identity, workspace, NovaHostPhase.failed, navigation, null, implemented)
        return NovaSessionHost(identity, workspace, NovaHostPhase.ready,
            navigation.updateAvailability(enabled.intersect(implemented), navigation.epoch), null, implemented)
    }

    fun fail(ticket: NovaAvailabilityTicket): NovaSessionHost = if (!matches(ticket)) this else
        NovaSessionHost(identity, workspace, NovaHostPhase.failed, navigation, null, implemented)

    fun apply(event: NovaNavigationEvent, from: String): NovaSessionHost = if (!isCurrent(from)) this else
        NovaSessionHost(identity, workspace, phase, navigation.apply(event, from), pending, implemented)

    fun isCurrent(epoch: String) = phase == NovaHostPhase.ready && epoch == navigation.epoch
    fun isCurrent(epoch: String, workspaceID: UUID, permissionRevision: Long) = isCurrent(epoch) &&
        workspace?.workspaceID == workspaceID && workspace.permissionRevision == permissionRevision

    // Same-epoch company/query/request races still require a feature-specific request guard.
    fun <Value> scope(value: Value, from: String): NovaScopedValue<Value>? =
        if (isCurrent(from)) NovaScopedValue(from, value) else null
    fun <Value> value(snapshot: NovaScopedValue<Value>?): Value? =
        if (snapshot != null && isCurrent(snapshot.epoch)) snapshot.value else null

    private fun matches(ticket: NovaAvailabilityTicket) = phase == NovaHostPhase.resolving && pending === ticket &&
        identity == ticket.identity && workspace == ticket.workspace && navigation.epoch == ticket.epoch
    private fun freshNavigation() = NovaNavigationState(UUID.randomUUID().toString(), emptySet())
}
